Setup Travis CI for build AppImage and deploy it to GitHub

Compile this application on [Travis CI](https://travis-ci.org/) upon each `git push`, and upload an [AppImage](http://appimage.org/) to your GitHub Releases page.

Providing an [AppImage](http://appimage.org/) would have, among others, these advantages:

- Applications packaged as an AppImage can run on many distributions (including Ubuntu, Fedora, openSUSE, CentOS, elementaryOS, Linux Mint, and others)
- One app = one file = super simple for users: just download one AppImage file, [make it executable](http://discourse.appimage.org/t/how-to-make-an-appimage-executable/80), and run
- No unpacking or installation necessary
- No root needed
- No system libraries changed
- Works out of the box, no installation of runtimes needed
- Optional desktop integration with `appimaged`
- Optional binary delta updates, e.g., for continuous builds (only download the binary diff) using AppImageUpdate
- Can optionally GPG2-sign your AppImages (inside the file)
- Works on Live ISOs
- Can use the same AppImages when dual-booting multiple distributions
- Can be listed in the [AppImageHub](https://appimage.github.io/apps) central directory of available AppImages
- Can double as a self-extracting compressed archive with the `--appimage-extract` parameter

[Here is an overview](https://appimage.github.io/apps) of projects that are already distributing upstream-provided, official AppImages.

### TODO

1. https://github.com/appimage/appimagekit
1.1 https://github.com/probonopd/linuxdeployqt
1.2 https://github.com/probonopd/uploadtool

**PLEASE NOTE:** For this to work, you need to enable Travis CI for your repository as [described here](https://travis-ci.org/getting_started) **prior to merging this**, if you haven't already done so. Also, You need to set up `GITHUB_TOKEN` in Travis CI for this to work; please see https://github.com/probonopd/uploadtool.

If you have questions, AppImage developers are on #AppImage on irc.freenode.net.

Started work at

https://github.com/probonopd/photoflare/blob/f3df3cb2430d7abbba50f2bfa1e32e93e13a5d77/.travis.yml

but getting a compile error:

```
src/managers/FilterManager.cpp: In member function ‘QImage FilterManager::fitImage(const QImage&)’:
src/managers/FilterManager.cpp:388:18: error: ‘class Magick::Image’ has no member named ‘thumbnail’
     magickImage->thumbnail(Magick::Geometry(640,400,0,0,false,false));
                  ^
src/managers/FilterManager.cpp: In member function ‘QImage FilterManager::cropToCenter(const QImage&)’:
src/managers/FilterManager.cpp:465:18: error: ‘class Magick::Image’ has no member named ‘extent’
     magickImage->extent(Magick::Geometry(640,400,0,0,false,false),Magick::Color(127,127,127),Magick::StaticGravity);
                  ^
src/managers/FilterManager.cpp: In member function ‘QPolygon FilterManager::selectArea(const QImage&, const QPoint&, int, bool)’:
src/managers/FilterManager.cpp:773:38: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for(int j=0; j<magickImage->rows(); j++)
                                      ^
src/managers/FilterManager.cpp:776:45: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
         for(int i=0; i<magickImage->columns(); i++)
                                             ^
src/managers/FilterManager.cpp:783:47: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
             if((i == magickImage->columns() - 1) && marked)
                                               ^
src/managers/FilterManager.cpp:786:75: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
             if(magickImage->pixelColor(i,j) == Magick::ColorRGB (0, 0, 0) && !marked ||
                                                                           ^
src/managers/FilterManager.cpp:801:47: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
             if((i == magickImage->columns() - 1) && marked)
                                               ^
src/managers/FilterManager.cpp:804:74: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
             if(magickImage->pixelColor(i,j) == Magick::ColorRGB(0, 0, 0) && !marked ||
                                                                          ^
src/managers/FilterManager.cpp: In member function ‘QImage FilterManager::floodFillOpacity(const QImage&, const QColor&, int)’:
src/managers/FilterManager.cpp:829:38: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for(int j=0; j<magickImage->rows(); j++)
                                      ^
src/managers/FilterManager.cpp:831:45: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
         for(int i=0; i<magickImage->columns(); i++)
                                             ^
src/managers/FilterManager.cpp: At global scope:
src/managers/FilterManager.cpp:593:28: warning: ‘MagickLib::ChannelType channelById(int)’ defined but not used [-Wunused-function]
 static Magick::ChannelType channelById(int channelId)
                            ^
make: *** [FilterManager.o] Error 1
make: *** Waiting for unfinished jobs....
```

https://travis-ci.org/probonopd/photoflare/builds/466290284#L886

--- a/.travis.yml
+++ b/.travis.yml
@@ -5,15 +5,44 @@ dist: trusty
 
 before_install:
     - sudo add-apt-repository ppa:beineri/opt-qt-5.10.1-trusty -y
+    - sudo add-apt-repository ppa:ubuntu-toolchain-r/test -y
     - sudo apt-get update
     
-install: 
-    - sudo apt-get -y install libc6 libgraphicsmagick++1-dev qt511base
+install:
+    - sudo apt-get -y install libc6 qt510base qt510tools gcc-8
     - source /opt/qt5*/bin/qt5*-env.sh
 
 script:
-  "/bin/sh -c cd /var/photoflare/ && ./build.sh"
+  - wget "http://ftp.icm.edu.pl/pub/unix/graphics/GraphicsMagick/1.3/GraphicsMagick-1.3.31.tar.xz"
+  - tar -xf GraphicsMagick-*.tar.xz
+  - cd GraphicsMagick-*/
+  - ./configure --prefix=/usr --enable-shared
+  - make -j$(nproc)
+  - sudo make install
+  - cd ..
+  - qmake CONFIG+=release PREFIX=/usr
+  - make -j$(nproc)
+  - make INSTALL_ROOT=appdir -j$(nproc) install ; find appdir/
+  - wget -c -nv "https://github.com/probonopd/linuxdeployqt/releases/download/continuous/linuxdeployqt-continuous-x86_64.AppImage"
+  - chmod a+x linuxdeployqt-continuous-x86_64.AppImage
+  - unset QTDIR; unset QT_PLUGIN_PATH ; unset LD_LIBRARY_PATH
+  - export VERSION=$(git rev-parse --short HEAD) # linuxdeployqt uses this for naming the file
+  - ./linuxdeployqt*.AppImage ./appdir/usr/share/applications/*.desktop -bundle-non-qt-libs
+  - # Workaround to increase compatibility with older systems; see https://github.com/darealshinji/AppImageKit-checkrt for details
+  - mkdir -p appdir/usr/optional/ ; wget -c https://github.com/darealshinji/AppImageKit-checkrt/releases/download/continuous/exec-x86_64.so -O ./appdir/usr/optional/exec.so
+  - mkdir -p appdir/usr/optional/libstdc++/ ; cp /usr/lib/x86_64-linux-gnu/libstdc++.so.6 ./appdir/usr/optional/libstdc++/
+  - ( cd appdir ; rm AppRun ; wget -c https://github.com/darealshinji/AppImageKit-checkrt/releases/download/continuous/AppRun-patched-x86_64 -O AppRun ; chmod a+x AppRun)
+  - ./linuxdeployqt*.AppImage --appimage-extract
+  - export PATH=$(readlink -f ./squashfs-root/usr/bin):$PATH
+  - ./squashfs-root/usr/bin/appimagetool -g ./appdir/ PhotoFlare-$VERSION-x86_64.AppImage
 
+after_success:
+  - find appdir -executable -type f -exec ldd {} \; | grep " => /usr" | cut -d " " -f 2-3 | sort | uniq
+  - # curl --upload-file PhotoFlare*.AppImage https://transfer.sh/PhotoFlare-git.$(git rev-parse --short HEAD)-x86_64.AppImage
+  - wget -c https://github.com/probonopd/uploadtool/raw/master/upload.sh
+  - bash upload.sh PhotoFlare*.AppImage*
+  
 branches:
-  only:
-    - master
+  except:
+    - # Do not build tags that we create when we upload to GitHub Releases
+    - /^(?i:continuous)/

--- a/installers/deb/DEBIAN/usr/share/applications/photoflare.desktop
+++ b/installers/deb/DEBIAN/usr/share/applications/photoflare.desktop
@@ -8,7 +8,6 @@ Icon=photoflare
 Terminal=false
 Type=Application
 Categories=Graphics;2DGraphics;
-Keywords=photo,image,picture,photography,paint,graphics,editor
+Keywords=photo;image;picture;photography;paint;graphics;editor;
 StartupNotify=false
 MimeType=image/bmp;image/gif;image/jpeg;image/jpg;image/pjpeg;image/png;image/svg+xml;image/tiff;image/x-bmp;image/x-gray;image/x-icb;image/x-ico;image/x-png;image/x-portable-anymap;image/x-portable-bitmap;image/x-portable-graymap;image/x-portable-pixmap;image/x-xbitmap;image/x-xpixmap;image/x-pcx;image/x-targa;image/x-tga;image/openraster;
-

